### DISCUSSION TITLE
Realtime: How to Get Near-Guaranteed Delivery?

### DISCUSSION THREAD

**Discussion by Marviel** [supabase/supabase#6697] [Questions]
The lack of guaranteed delivery via this library makes me nervous, so I'm trying to figure out what I'll be able to squeeze out of it.

Assuming one wanted to get near-guaranteed delivery of all messages (say ~99.999% message delivery), what steps would one take in the Supabase infrastructure to do this?

I'm aware that the two primary points mentioned are:
- Ensure Postgres WAL has enough Memory / Storage.
- Ensure Realtime server has enough Memory / Storage.

I'm not sure if these variables are easily tweakable via the hosted instance of Supabase, or not.

If self-hosting, it would be interesting to be able to gauge exactly when these variables should be scaled.

Does anyone have advice on how to ensure higher reliability guarantees for Realtime messages?

**Comment by w3b6x9** [comment]
@Marviel as Realtime currently stands there are no guarantees even if you maxed out resources in either hosted or self-hosted Realtime versions. It wasn't a requirement when Realtime was conceived and since conception we haven't prioritized it. We have discussed it internally but no plans at the moment to add it.

**Reply by brianmutiso** [reply]
that's total crap, based on your reasoning when is delivery not important?

**Reply by alita-moore** [reply]
I don't see how we can use this in production without guaranteed delivery

**Comment by Marviel** [comment]
That's totally fair -- maybe let me flip the question around -- if you were
to conjecture, under what circumstances one would assume delivery would
*not* be guaranteed, based on your current understanding?

**Reply by w3b6x9** [reply]
@Marviel apologies on the late reply.

circumstances include when clients have intermittent internet connection, when Realtime temporarily cannot connect to the database (replication slot is temporary so there's no way to continue where Realtime left off when reconnecting however this is mitigated by making the slot permanent).

**Comment by alita-moore** [comment]
It seems strange to me that Supabase doesn't have some sort of functionality for the client that will log receipts for messages and then resend ones it fails to send. I can do this pretty easily using a series of tables, triggers, etc. But I'd of course prefer that I not have to manage that directly. Is this a thing that others do or is it generally not necessary?

**Reply by GaryAustin1** [reply]
I detect any subscription error and if the browser tab is visible, or when it becomes visible I reload form the monitored table any data I may have missed.  That could be by a timestamp on records or just reload the table entirely depending on the data.
